Egress firewall support

danwinship opened the pull request:
The origin side of https://github.com/openshift/openshift-sdn/pull/319, for reference. The interesting bits are all over there though.

dcbw commented:
Travis says you need a gofmt on pkg/sdn/api/v1beta3/types.go, pkg/sdn/api/v1/types.go and pkg/sdn/api/validation/validation_test.go

dcbw commented:
Otherwise LGTM but somebody with more experience adding API to origin needs to check too.

marun commented on pkg/client/egressfirewall.go:
```diff
@@ -0,0 +1,50 @@
+// Get returns information about a particular firewall
+func (c *egressFirewall) Get(name string) (result *sdnapi.EgressFirewall, err error) {
+	result = &sdnapi.EgressFirewall{}
+	err = c.r.Get().Resource("egressFirewalls").Name(name).Do().Into(result)
```
I'm reading the kube api conventions doc [1] and at the end of the opening it mentions that collections should be lowercase and plural (e.g. egressfirewalls instead of egressFirewalls).  Do you know if origin follows the same convention? 

1: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md

danwinship commented:
merged in the old openshift-sdn PR, and updated to add service endpoint filtering; we use the existing endpoint filtering infrastructure to also filter out external endpoints that conflict with firewall rules. (This is a stopgap; eventually we will fix things so that either service rewriting happens in OVS [when we have conntrack support there] or else egress firewalling happens in iptables [when we have a way to make iptables rules namespace-aware].)

danwinship commented:
@openshift/networking PTAL?

The one customer who has looked at this so far was a bit concerned about having all of the rules on a single object, since they expect to have on the order of one rule per namespace (with lots of namespaces). (Basically, they plan to block all outgoing access by default, and project admins will have to use some [non-OpenShift] tool to indicate the external sites they need access to, and then some script will create the EgressFirewall rules from that.) So, we could change it so that the EgressFirewall object doesn't have a namespace field, and instead you just create objects in each namespace, and then say that the default namespace's rules apply to all namespaces, except that as discussed before in email, I'm dubious of that since nothing else treats the default namespace as being a parent of other namespaces. So maybe we should just ignore this for 3.3 and just make sure it gets dealt with better in the eventual upstream version.

danwinship commented:
@openshift/api-review PTAL. This adds a new object type `EgressFirewall` containing namespaced firewall rules for the cluster as a whole. (It does not change any existing APIs.) See https://trello.com/c/iH5IjWI8 for the use cases.

smarterclayton commented:
How does this differ from NetworkPolicy in intent?

On Jul 14, 2016, at 9:47 AM, Dan Winship <notifications@github.com> wrote:

@openshift/api-review <https://github.com/orgs/openshift/teams/api-review>
PTAL. This adds a new object type EgressFirewall containing namespaced
firewall rules for the cluster as a whole. (It does not change any existing
APIs.) See https://trello.com/c/iH5IjWI8 for the use cases.

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9227#issuecomment-232670143>, or mute
the thread
<https://github.com/notifications/unsubscribe/ABG_py0ren0-4k6QxX9lJyfK4PKlOH1Iks5qVj3ygaJpZM4IxESt>
.

danwinship commented:
As of kubernetes 1.3, NetworkPolicy only covers ingress policy, not egress policy. There will eventually be upstream egress policy stuff as well, but not in time for OpenShift 1.3 (or probably even 1.4), and based on the conversation that has happened so far, it's impossible to guess what it will look like when it does happen. (Notably, the existing ingress NetworkPolicy stuff is something that project admins opt in to, whereas the functionality implemented in this PR is imposed by the cluster admin, and can't be overridden or modified by project admins, so it seems unlikely to end up being part of the same NetworkPolicy objects.) So the expectation is that, like NetworkNamespaces, this would eventually be deprecated in favor of the upstream solution, but we don't know what or when that will be.

smarterclayton commented:
Why would cluster admins not be able to choose whether either ingress *or*
egress policy is applied?  Can you describe the use cases where a project
admin would be allowed to set ingress but not egress and vice versa (I just
want to make sure that we need two objects, vs storing this as an
annotation on network policy and making it virtual).

On Thu, Jul 14, 2016 at 10:49 AM, Dan Winship <notifications@github.com>
wrote:

> As of kubernetes 1.3, NetworkPolicy only covers ingress policy, not egress
> policy. There will eventually be upstream egress policy stuff as well, but
> not in time for OpenShift 1.3 (or probably even 1.4), and based on the
> conversation that has happened so far, it's impossible to guess what it
> will look like when it does happen. (Notably, the existing ingress
> NetworkPolicy stuff is something that project admins opt in to, whereas the
> functionality implemented in this PR is imposed by the cluster admin, and
> can't be overridden or modified by project admins, so it seems unlikely to
> end up being part of the same NetworkPolicy objects.) So the expectation is
> that, like NetworkNamespaces, this would eventually be deprecated in favor
> of the upstream solution, but we don't know what or when that will be.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9227#issuecomment-232687855>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_pxT7zyw9IPv4BoYMJ6f5U7ogR51fks5qVkxrgaJpZM4IxESt>
> .
>

smarterclayton commented on pkg/sdn/api/types.go:
```diff
@@ -60,3 +60,33 @@ type NetNamespaceList struct {
+	unversioned.TypeMeta
+	kapi.ObjectMeta
+
+	Rules []EgressFirewallRule
```
Should be in a spec (same as network policy) and Rules should be called `Egress []EgressFirewallRule`

smarterclayton commented on pkg/sdn/api/types.go:
```diff
@@ -60,3 +60,33 @@ type NetNamespaceList struct {
+}
+
+// EgressFirewall provides a list of firewall rules on outgoing traffic
+type EgressFirewall struct {
```
Should be called `EgressNetworkPolicy`

smarterclayton commented on pkg/sdn/api/types.go:
```diff
@@ -60,3 +60,33 @@ type NetNamespaceList struct {
+// EgressFirewallRule contains a single outgoing traffic firewall rule
+type EgressFirewallRule struct {
+	Policy EgressFirewallPolicy
+	From   string // Namespace name or empty
```
Should be a struct:

```
From EgressPolicyPeer
...

type EgressPolicyPeel struct {
  Namespace string
}
```

danwinship replied:
To be consistent with the upstream proposals so far, we would make EgressPolicyPeer be the `To` field, not the `From` field.

And to be consistent with the upstream proposals, `From` would not exist at all: instead, we would allow creating `EgressFirewall`/`EgressNetworkPolicy`/whatever objects in any namespace, and each object's rules would apply to traffic coming from the namespace that that object is in. (They would have to still be only creatable by cluster admins though, not project admins.)

We would still need somewhere to specify the final fallback/override rules for all remaining traffic (from all namespaces) that wasn't explicitly accepted or denied by any previous rule. None of the upstream proposals so far have attempted to handle that though, so there's no precedent to follow. We could use the default namespace for that, which kind of makes sense (and kind of doesn't). ?

smarterclayton replied:
On Jul 15, 2016, at 10:14 AM, Dan Winship <notifications@github.com> wrote:

In pkg/sdn/api/types.go
<https://github.com/openshift/origin/pull/9227#discussion_r70979582>:

> @@ -60,3 +60,33 @@ type NetNamespaceList struct {
>  	unversioned.ListMeta
>  	Items []NetNamespace
>  }
> +
> +// EgressFirewallPolicy gives the type of an EgressFirewallRule
> +type EgressFirewallPolicy string
> +
> +const (
> +	EgressFirewallPolicyAllow EgressFirewallPolicy = "allow"
> +	EgressFirewallPolicyDeny  EgressFirewallPolicy = "deny"
> +)
> +
> +// EgressFirewallRule contains a single outgoing traffic firewall rule
> +type EgressFirewallRule struct {
> +	Policy EgressFirewallPolicy
> +	From   string // Namespace name or empty

To be consistent with the upstream proposals so far, we would make
EgressPolicyPeer be the To field, not the From field.

And to be consistent with the upstream proposals, From would not exist at
all: instead, we would allow creating
EgressFirewall/EgressNetworkPolicy/whatever
objects in any namespace, and each object's rules would apply to traffic
coming from the namespace that that object is in. (They would have to still
be only creatable by cluster admins though, not project admins.)



That's generally what we do anyway - quota and limit apply to their
namespace, and they are only editable by cluster admin.


We would still need somewhere to specify the final fallback/override rules
for all remaining traffic (from all namespaces) that wasn't explicitly
accepted or denied by any previous rule. None of the upstream proposals so
far have attempted to handle that though, so there's no precedent to
follow. We could use the default namespace for that, which kind of makes
sense (and kind of doesn't). ?


Do we need fallback rules?  For quota we simply distribute the rules to all
namespaces.  For other places we create cluster scoped resources or use
global config.

If we think in the short term we don't need project admins editing egress
rules, then let's just have no fallback, and require admins to specify them
(we already effectively do this for quota, so again, no different).


—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9227/files/7d5bf0193c381259ef2b0bc4170d5cc76a113231#r70979582>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_pzu2EBxB6RFg8H4bJHcDs8tXMdb7ks5qV5XWgaJpZM4IxESt>
.

danwinship replied:
> Do we need fallback rules? For quota we simply distribute the rules to all namespaces. For other places we create cluster scoped resources or use global config.

Can you point me to an example of where we have both namespace-scoped and cluster-scoped versions of a resource? As I understand it, this would require a second resource type (`ClusterEgressFirewall` or whatever) since a given resource type has to be either always namespaced or never namespaced, right?

(One possibility would be to add an array of `EgressFirewallRule` to the `ClusterNetwork` record?)

> If we think in the short term we don't need project admins editing egress rules, then let's just have no fallback, and require admins to specify them (we already effectively do this for quota, so again, no different).

Hm... I think if we were going to do this we'd have to at least add a global "deny egress by default" option, so that if the admin forgets to set up the firewall for a new project, then the project would default to being completely locked down rather than completely open.

smarterclayton replied:
We don't do that for quota, which is the same outcome.  All users
should be using new-project (create project directly is admin), which
uses the template, so admins can't forget.

smarterclayton commented on pkg/sdn/api/validation/validation.go:
```diff
@@ -117,3 +117,35 @@ func ValidateNetNamespace(netnamespace *sdnapi.NetNamespace) field.ErrorList {
+	allErrs := validation.ValidateObjectMeta(&egressFirewall.ObjectMeta, false, oapi.MinimalNameRequirements, field.NewPath("metadata"))
+
+	if egressFirewall.Name != "default" {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("name"), egressFirewall.Name, "EgressFirewall name must be 'default'"))
```
Please don't assume that there is only one egress firewall rule in the API.  That breaks other API conventions and assumptions.  If you only honor default from the SDN client, that's fine, but don't make APIs that are "fake" singletons.

smarterclayton commented:
Added some comments about structure assuming this is intended to be roughly in line with the upstream NetworkPolicy.  Unless you plan on being wildly divergent, we should have similar patterns and structures.

danwinship commented:
> Why would cluster admins not be able to choose whether either ingress *or* egress policy is applied?

Because ingress filtering is about protecting your project from other people, but egress filtering is about protecting other people from your project. The entire premise of egress filtering (in this PR) is that the cluster admin doesn't fully trust the project admins. (Eg, in OpenShift Online, maybe there are some
monitoring/cleanup/whatever jobs that run in pods in the cluster but need to have access to some internal Red Hat logging servers, yum repos, etc. But we don't want customer pods to be able to access those servers. So, we create an egress firewall blocking access to those servers, and add an exception for the namespace that the special jobs run in.)

There are other egress filtering use cases where opt-in at the project level does make sense. (Eg, "I know my pod needs to contact github.com, but nothing else, so I want to firewall off everything else so that if hackers break into my pod they can't use it to attack people") But this PR isn't trying to handle that. (And really, that makes more sense as a Pod annotation than a Namespace-level policy...)

danwinship commented:
> Added some comments about structure assuming this is intended to be roughly in line with the upstream NetworkPolicy.

It's not. This isn't intended to handle arbitrary "egress policy". It's just supposed to handle this one use case (namespace-aware firewall rules) as a stop-gap until upstream finishes designing a full egress policy solution, at which point we should hopefully be able to to auto-translate EgressFirewall objects into corresponding upstream policy objects and deprecate EgressFirewall.

Also, I don't think it's safe to assume at this point that the eventual upstream egress filtering solution is going to look anything like (current) NetworkPolicy. (The initial attempt to add egress policy to NetworkPolicy fell apart when someone pointed out that the proposal didn't actually let you implement any of the proposed use cases.)

smarterclayton commented:
Rather than invent new conventions, use the existing conventions.  If
you're going to deprecate the object, that's fine, but we don't add APIs
that are inconsistent.  I don't care whether eventual egress network policy
matches what we do here, but don't create something net new that doesn't
look like anything that has gone before.

On Thu, Jul 14, 2016 at 1:23 PM, Dan Winship <notifications@github.com>
wrote:

> Added some comments about structure assuming this is intended to be
> roughly in line with the upstream NetworkPolicy.
>
> It's not. This isn't intended to handle arbitrary "egress policy". It's
> just supposed to handle this one use case (namespace-aware firewall rules)
> as a stop-gap until upstream finishes designing a full egress policy
> solution, at which point we should hopefully be able to to auto-translate
> EgressFirewall objects into corresponding upstream policy objects and
> deprecate EgressFirewall.
>
> Also, I don't think it's safe to assume at this point that the eventual
> upstream egress filtering solution is going to look anything like (current)
> NetworkPolicy. (The initial attempt to add egress policy to NetworkPolicy
> fell apart when someone pointed out that the proposal didn't actually let
> you implement any of the proposed use cases.)
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9227#issuecomment-232733016>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_pwKBDeG7R-s0UubUEKUxHd8-dIHMks5qVnB6gaJpZM4IxESt>
> .
>

danwinship commented:
[test]

danwinship commented:
> Rules or EgressRules. Egress is redundant probably - you're not going to add ingress rules to an egress policy.

I had it as Rules originally but you had requested I change it to Egress for consistency with NetworkPolicySpec, which has `Ingress []NetworkPolicyIngressRule`. I'm fine with either name...

> I'm still hung up on this being EgressFirewall instead of EgressNetworkPolicy. Can you articulate why Firewall is better than *NetworkPolicy? (we are using *Policy in a lot more places now, so the burden is to justify the difference)

I had stuck with EgressFirewall because this probably isn't going to cover all of the use cases that the upstream egress network policy does, so I figured a more limited name made sense. But I can switch it to EgressNetworkPolicy.

smarterclayton commented:
Ok, been long enough that I forgot.  Leave egress alone (ingress makes
sense on upstream because the total object is NetworkPolicy).

If we had more use cases in the short term, would we add them to our
object or wait for upstream egress policy?

danwinship commented:
- Changed "allow"/"deny" constants to "Allow"/"Deny"
- Renamed EgressFirewall to EgressNetworkPolicy, with associated other renamings. In particular, renamed "Policy EgressFirewallPolicy" ("allow"/"deny") to "Type EgressNetworkPolicyRuleType"
- Redid describer with single-line rules
- Moved rule counting from plugin to validator

smarterclayton commented on pkg/sdn/api/validation/validation.go:
```diff
@@ -117,3 +117,29 @@ func ValidateNetNamespace(netnamespace *sdnapi.NetNamespace) field.ErrorList {
+	}
+
+	if len(policy.Spec.Egress) > 50 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("egress"), "", ("too many egress rules")))
```
Include the maximum in the message so the user doesn't have to guess.

smarterclayton commented:
Is there any review needed for the SDN changes?  API is approved, I don't see anything obvious but I am not a OVS expert.

smarterclayton commented on pkg/sdn/api/types.go:
```diff
@@ -60,3 +60,42 @@ type NetNamespaceList struct {
+	Egress []EgressNetworkPolicyRule
+}
+
+// EgressNetworkPolicy describes the current egress network policy
```
Flush out this godoc - preferably 3-4 sentences.  End user needs to be able to figure out what this object does and how it is used from the swagger doc.

smarterclayton replied:
... on the v1 object (not here)

smarterclayton commented on pkg/sdn/api/v1/types.go:
```diff
@@ -80,3 +80,50 @@ type NetNamespaceList struct {
+
+// EgressNetworkPolicyPeer specifies a target to apply egress policy to
+type EgressNetworkPolicyPeer struct {
+	// CIDRSelector is the CIDR range to allow/deny traffic to
```
Use the external values in the godoc on public API fields so that swagger has the correct name (someone seeing swagger doesn't use go, they use JSON), so for all of these `cidrSelector is the CIDR range`

danwinship commented:
Pushed docs update (including validation error message fix). I'll get someone on @openshift/networking to review the rest and merge.

smarterclayton commented:
Thanks, API approved.

danwinship commented:
Hm... ok, problem: the implementation of the egress fireall is based on NetNamespaces, but the rules are specified per-(regular)-Namespace... so what happens if you "oadm join-projects" two Namespaces that each have their own EgressNetworkPolicy object into a single NetNamespace?

I think the answer to that is going to have to be that all namespaces sharing a NetNamespace share all of the egress firewall rules too, and in cases where there are multiple EgressNetworkPolicy objects applying to a NetNamespace, it is undefined what order they get applied in (and so either, you better only have one, or else they shouldn't conflict.)

That would also let us get rid of the requirement that we only look at the EgressNetworkPolicy object named "default"; instead, we just say that all EgressNetworkPolicy objects get applied, but in an undefined order. Depending on the kind of rules you're using, this could even be a useful feature. And it's more like upstream NetworkPolicy...

@dcbw does this make sense to you?

knobunc commented:
@danwinship that behavior makes sense.  It would be nice if we could predict an ordering for debugging these things.

danwinship commented:
If you can predict the order it's not really "undefined"... the point is more that you shouldn't have multiple policies unless they're written in such a way that the ordering doesn't matter. (ie, don't put "allow X" in one object and "deny Y" in another if X and Y overlap.)

danwinship commented:
(ie, we shouldn't ever be "debugging these things", because people shouldn't be doing those things)

pecameron commented:
On 07/25/2016 11:14 AM, Dan Winship wrote:
>
> (ie, we shouldn't ever be "debugging these things", because people 
> shouldn't be doing those things)
>
However, test cases should...
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/openshift/origin/pull/9227#issuecomment-234983690>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ANUgep2c6m0ha0GjCJk8qZbOII5VASb1ks5qZNLvgaJpZM4IxESt>.
>

dcbw commented:
> Hm... ok, problem: the implementation of the egress fireall is based on NetNamespaces, but the rules are specified per-(regular)-Namespace... so what happens if you "oadm pod-network join-projects" two Namespaces that each have their own EgressNetworkPolicy object into a single NetNamespace?

@danwinship What I'd like to see happen is that each namespace gets its own VNID that never changes.  Join/Split would simply add flow rules that would allow traffic between those two VNIDs instead of changing the VNID itself.  This shouldn't hugely increase the number of flow rules because I expect the number of joined projects to not be large; but it may be a larger issue for OpenShift Online.

This would also allow us to explore using the VNID as part of the IP address if we wanted to experiment with that, since pod IP addresses have to be stable for the life of the pod.

If we made VNIDs stable, then we don't really need NetNamespaces since they were only a construct to allow two namespaces to share the same VNID.  And thus each Namespace could still retain its own egress policy *and* be joined to another namespace without conflicts.

> I think the answer to that is going to have to be that all namespaces sharing a NetNamespace share all of the egress firewall rules too, and in cases where there are multiple EgressNetworkPolicy objects applying to a NetNamespace, it is undefined what order they get applied in (and so either, you better only have one, or else they shouldn't conflict.)

Yes, I think that's the best option for now.

> That would also let us get rid of the requirement that we only look at the EgressNetworkPolicy object named "default"; instead, we just say that all EgressNetworkPolicy objects get applied, but in an undefined order. Depending on the kind of rules you're using, this could even be a useful feature. And it's more like upstream NetworkPolicy...

Yeah, though I'd like to go in the direction of stable VNIDs and take the extra OVS flow rule hit, which @pecameron has benchmarked as not being a huge problem.  Plus I think we can optimize that somewhat if we need to.  Thoughts on that as a long-term path?

danwinship commented:
@pravisankar accidentally reviewed this branch on #9982 instead (that branch being built on top of this one). Replying to his comments here:

> Enable this resource only if openshift-sdn plugin is used?

That's inconsistent with the other resource types (eg, ClusterNetwork, HostSubnet) which are also openshift-sdn-specific, but exist regardless (and we discussed this in another PR recently and agreed that the resources should always exist).

> We are limiting number of egress rules to avoid/reduce network performance degradation?

Not really, although yes. OVS doesn't have implicit rule ordering, so each flow needs a distinct priority value, so the total number of egress rules (per namespace) has to be less than the total number of possible priority values or we can't generate the flows correctly. The total number of priority values is 65536, but no one needs that many rules, so Clayton said to start small, and we could bump it up later if needed.

> Stale comment or something needs to be done here?

(Re "`// FIXME: poke the endpoint-syncer somehow...`"). Ideally something would be done, but there's currently nothing we can do. So this means that if a namespace already has a service pointing to 1.2.3.4, and then you add a firewall rule blocking access to 1.2.3.4, the service endpoint won't be dropped until the next full endpoint resync, 0-30 seconds later.

> Where are we creating the 'default' egress policy?

We aren't, the code was just requiring that the admin-created egress policy had to be called that. But the next version of the branch drops that restriction and instead just requires that there only be one EgressNetworkPolicy in the namespace, but with any name you want.

> This has some side effects.

Yes, the other PR was based on an older version of the branch, and the Namespace vs NetNamespace interactions were discussed here. That will all be addressed in the next version of the branch

danwinship commented:
Updated based on discussion yesterday: now we disallow EgressNetworkPolicy in any Namespace that shares a vnid with any other Namespace. This was the solution that seemed most compatible with using NetNamespace to implement it now (in particular, this means there's no situation where traffic in one Namespace is affected by an EgressNetworkPolicy object in another Namespace) without causing migration headaches if we move to vnid-per-Namespace / deprecation of NetNamespace in the future (any setup that is legal today will still have the same behavior in that future, and additional setups can become legal at that point).

@dcbw @pravisankar PTAL

(I know `vnids_test.go` will probably end up going away soon, but I didn't trust the new code without it having a test case...)

danwinship commented:
(Oh, and I'm aware that proxy.go and controller.go behave inconsistently in the EgressNetworkPolicy-in-shared-NetNamespace case, but fixing that would require making the proxy also track VNIDs, and since we're saying that situation is not allowed anyway, it didn't make sense to do that.)

danwinship commented:
[test]

pravisankar commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -957,8 +957,13 @@ func printClusterNetworkList(list *sdnapi.ClusterNetworkList, w io.Writer, opts
+	if _, err := fmt.Fprintf(w, "%s\t", n.Namespace); err != nil {
+		return err
+	}
+	if _, err := fmt.Fprintf(w, "%s\n", n.Name); err != nil {
```
Simplify...if _, err := fmt.Fprintf(w, "%s\t%s\n", n.Namespace, n.Name); err != nil
egressNetworkPolicyColumns contains only one column 'NAME', do we want to add 'NAMESPACE' column?

danwinship replied:
Hm... somehow I thought this matched the style elsewhere, but I guess not really.
Namespace gets added automatically, although it looks like I didn't handle it quite correctly.

(Also, this was a fix-up that I squashed into the wrong commit, it should have gone into the earlier commit...)

pravisankar commented on pkg/sdn/plugin/controller.go:
```diff
@@ -330,6 +337,155 @@ func (plugin *OsdnNode) SetupSDN(localSubnetCIDR, clusterNetworkCIDR, servicesNe
+
+		vnid, err := plugin.vnids.GetVNID(policy.Namespace)
+		if err != nil {
+			glog.Warningf("Could not find namespace %q for EgressNetworkPolicy: %v", err)
```
glog.Warningf("Could not find netid for namespace %q: %v", policy.Namespace, err)

pravisankar commented on pkg/sdn/plugin/controller.go:
```diff
@@ -330,6 +337,155 @@ func (plugin *OsdnNode) SetupSDN(localSubnetCIDR, clusterNetworkCIDR, servicesNe
+	for _, policy := range policies {
+		vnid, err := plugin.vnids.GetVNID(policy.Namespace)
+		if err != nil {
+			glog.Warningf("Could not find namespace %q for EgressNetworkPolicy: %v", err)
```
glog.Warningf("Could not find netid for namespace %q: %v", policy.Namespace, err)

pravisankar commented on pkg/sdn/plugin/controller.go:
```diff
@@ -330,6 +337,155 @@ func (plugin *OsdnNode) SetupSDN(localSubnetCIDR, clusterNetworkCIDR, servicesNe
+	} else if vnid == 0 {
+		glog.Errorf("EgressNetworkPolicy in global network namespace is not allowed (%s); ignoring", policyNames(policies))
+	} else if len(namespaces) > 1 {
+		glog.Errorf("EgressNetworkPolicy not allowed in shared NetNamespace (%s); dropping all traffic", strings.Join(namespaces, ", "))
```
These errors will be buried in the logs and the user will not know why the egress policy is not working. Need to provide some feedback mechanism to the user. Either sending events or warnings during egress policy creation.

pravisankar commented on pkg/sdn/plugin/controller.go:
```diff
@@ -330,6 +337,155 @@ func (plugin *OsdnNode) SetupSDN(localSubnetCIDR, clusterNetworkCIDR, servicesNe
+		otx.DeleteFlows("table=9, reg0=%d", vnid)
+		otx.AddFlow("table=9, reg0=%d, priority=1, actions=drop", vnid)
+	} else if len(policies) > 1 {
+		glog.Errorf("multiple EgressNetworkPolicies in same network namespace (%s) is not allowed; dropping all traffic", policyNames(policies))
```
Same comment as above

pravisankar commented on pkg/sdn/api/validation/validation.go:
```diff
@@ -117,3 +117,29 @@ func ValidateNetNamespace(netnamespace *sdnapi.NetNamespace) field.ErrorList {
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("egress").Index(i).Child("to"), rule.To.CIDRSelector, err.Error()))
+		}
+	}
```
Check for conflicting rules within a given policy?

danwinship replied:
"conflicting" how exactly? The rules are ordered, and it's expected that they may refer to overlapping ranges. (eg, "allow 192.168.1.0/24, deny 192.168.0.0/16") I guess we could check for impossible-to-match rules (like if those two were in the opposite order), but I don't think that kind of error is especially more likely than any other rule "typos" which we wouldn't be able to check ("oops, I meant 192.168.*2*.0") so I don't think it's really worth it.

pravisankar commented on pkg/sdn/api/validation/validation.go:
```diff
@@ -117,3 +117,29 @@ func ValidateNetNamespace(netnamespace *sdnapi.NetNamespace) field.ErrorList {
+		}
+	}
+
+	if len(policy.Spec.Egress) > 50 {
```
const EgressMaxAllowedRules = 50 in pkg/sdn/api/v1/types.go will be neat

pravisankar commented on pkg/sdn/plugin/controller.go:
```diff
@@ -330,6 +337,155 @@ func (plugin *OsdnNode) SetupSDN(localSubnetCIDR, clusterNetworkCIDR, servicesNe
+		// Temporarily drop all outgoing traffic, to avoid race conditions while modifying the other rules
+		otx.AddFlow("table=9, reg0=%d, cookie=1, priority=65535, actions=drop", vnid)
+		otx.DeleteFlows("table=9, reg0=%d, cookie=0/1", vnid)
+
```
Don't we need to drop all existing rules for vnid before adding new rules?
otx.DeleteFlows("table=9, reg0=%d", vnid) [Consider the case where policy is updated]

danwinship replied:
we do; the comment about dropping all traffic only applies to the AddFlow(); the DeleteFlows() then drops all existing flows for vnid (except for the "drop all traffic" flow which we just added with cookie=1 to mark it as being different from the the other flows)

pravisankar commented:
LGTM, there will be follow up PR to provide feedback to the user when egress policy is not applied.

pravisankar commented:
[test]

pravisankar commented:
Flake https://github.com/openshift/origin/issues/9959
[test]

pravisankar commented:
[merge]

dcbw commented:
LGTM re[merge]

smarterclayton commented:
[test] #9959

knobunc commented:
re[merge]

stevekuznetsov commented:
re[test]
